### ISSUE TITLE
Replace class init checks for failed initializations with exception throws

### ISSUE THREAD

**Issue by dmlloyd** [qbicc/qbicc#658]
Follow up to #657. If the interpreter fails to initialize a class, any class init checks for that class should be replaced with a throw of a new exception-in-initializer error.

This requires that there be a way to query the initialization status of each class.

**Comment by ashu-mehra** [comment]
If the build time initialization fails for a class, can't we postpone its initialization to run time?

**Comment by dmlloyd** [comment]
When it comes to compilation, we really don't want to rely on heuristics to determine the semantics of the program. Deferring initialization has a cascading effect as every class which depends on the class in question must then be deferred.

It is my thinking that the goal we should be working towards is having most or all initialization happening at build time, because almost every benefit of native images that are enjoyed today are derived from build time initialization. Run time initialization should probably be implemented using a separate mechanism.

**Comment by DanHeidinga** [comment]
> It is my thinking that the goal we should be working towards is having most or all initialization happening at build time, because almost every benefit of native images that are enjoyed today are derived from build time initialization. Run time initialization should probably be implemented using a separate mechanism.

I think this is too broad of a statement.  There are lots of things that can't - and shouldn't - be initialized at buildtime.  Some of these can be addressed by changes to JDK's class libraries or user code to delay the initialization but others can't be.  

Even in our discussion on class initialization [1], we recognized that some things need to be done at runtime and others may need to be "redone" at runtime.

Aiming for 100% BTI is a good prototype goal to flesh out the issues and help identify how we can work around them, but we should be clear that it's not a realistic end goal.

[1] https://github.com/qbicc/qbicc/discussions/523

**Comment by dmlloyd** [comment]
> > It is my thinking that the goal we should be working towards is having most or all initialization happening at build time, because almost every benefit of native images that are enjoyed today are derived from build time initialization. Run time initialization should probably be implemented using a separate mechanism.
> 
> I think this is too broad of a statement. There are lots of things that can't - and shouldn't - be initialized at buildtime. Some of these can be addressed by changes to JDK's class libraries or user code to delay the initialization but others can't be.

I should clarify that I didn't mean initializing things in general, I'm talking about initializing classes in particular. In my estimation, anything that needs to be done or redone at run time should be using a different mechanism than class initialization to make it happen. Otherwise we end up with the GraalVM problem where a run-time-initialized class infects its dependent tree since the JVM specification is fairly clear that classes cannot really be used until they are initialized.  Being unable to even _refer_ to a run-time-initialized class during build time has been a practical problem many times for us.

**Comment by DanHeidinga** [comment]
> > > It is my thinking that the goal we should be working towards is having most or all initialization happening at build time, because almost every benefit of native images that are enjoyed today are derived from build time initialization. Run time initialization should probably be implemented using a separate mechanism.
> > 
> > 
> > I think this is too broad of a statement. There are lots of things that can't - and shouldn't - be initialized at buildtime. Some of these can be addressed by changes to JDK's class libraries or user code to delay the initialization but others can't be.
> 
> I should clarify that I didn't mean initializing things in general, I'm talking about initializing classes in particular.

To be clear, so was I.  `<clinit>` is currently a mish-mash of state where some of it can be early initialized without issue while other bits (often in the same class!) can't be.

> In my estimation, anything that needs to be done or redone at run time should be using a different mechanism than class initialization to make it happen. 

The challenge is that that isn't the language we have today.  The "different mechanism" doesn't (yet) exist.  And there are real cases that shouldn't be BTI that need to be expressed to make existing programs run in this environment.

> Otherwise we end up with the GraalVM problem where a run-time-initialized class infects its dependent tree since the JVM specification is fairly clear that classes cannot really be used until they are initialized. Being unable to even _refer_ to a run-time-initialized class during build time has been a practical problem many times for us.

A class can be loaded but not initialized so I'm unclear on the "Being unable to even _refer_ to a run-time-initialized class during build time" problem.  

The case we've run into that I'm aware of is using special constants to indicate the platform / target environment we're building for - and I agree that's been a nuisance.  A special case of treating those classes, assuming it's a small number,  as "must be BTI" is certainly something we can do.  The general case feels too far for the existing applicationos we're starting with.

**Comment by dmlloyd** [comment]
> > > > It is my thinking that the goal we should be working towards is having most or all initialization happening at build time, because almost every benefit of native images that are enjoyed today are derived from build time initialization. Run time initialization should probably be implemented using a separate mechanism.
> > > 
> > > 
> > > I think this is too broad of a statement. There are lots of things that can't - and shouldn't - be initialized at buildtime. Some of these can be addressed by changes to JDK's class libraries or user code to delay the initialization but others can't be.
> > 
> > 
> > I should clarify that I didn't mean initializing things in general, I'm talking about initializing classes in particular.
> 
> To be clear, so was I. `<clinit>` is currently a mish-mash of state where some of it can be early initialized without issue while other bits (often in the same class!) can't be.
> 
> The challenge is that that isn't the language we have today. The "different mechanism" doesn't (yet) exist. And there are real cases that shouldn't be BTI that need to be expressed to make existing programs run in this environment.

It is not a requirement that we (or Leyden) need to try to make any existing software (including the JDK) be native-ready to _any_ degree on day one. In fact the Leyden project calls this out specifically. Given this, and due to the infectious nature of mixed class init, and given that almost every native image benefit is derived directly or indirectly from front-loading class initialization to build time, _and_ based on our experiences with Quarkus, I don't think it is nearly as expedient or useful as it would appear to be to compromise on this for the sake of compatibility with existing code. Artificially striving for any level compatibility (or for that matter introducing any other ancillary requirements of this kind) can easily negatively impact the quality of the end result (for example by potentially complicating the definition of "class initialization" in the JVM spec). It's just not as nearly important to be maximally compatible as it is for native images (or the broader "static images" given in the Leyden proposal) to justify their limitations by maximizing material benefit compared to a dynamic JVM in terms of image size and startup speed.

In the short term, using pragmatic/hacky substitutions would be more than adequate to bridge between worlds as we've shown already with Quarkus and as has been shown by the efforts of upstream library developers and users. It is certainly the case that pre-standardization, our mechanism and GraalVM's will not likely be compatible unless we went to extraordinary measures to make it so, which in context seems like a waste of effort that could be spent on pushing forward a standardizable solution.

In the medium to long term, libraries would be migrating towards the standard approach to having native-friendly run time initialization (which for most libraries, even specialized ones like Netty, would likely only be a quite small overall percentage of code). This migration would be helped along by community contributions (certainly including substantial contributions from us).

> A class can be loaded but not initialized so I'm unclear on the "Being unable to even _refer_ to a run-time-initialized class during build time" problem.

> The case we've run into that I'm aware of is using special constants to indicate the platform / target environment we're building for - and I agree that's been a nuisance. A special case of treating those classes, assuming it's a small number, as "must be BTI" is certainly something we can do. The general case feels too far for the existing applications we're starting with.

The applications we are starting with, beyond "hello world" of course, are the set of applications which Quarkus builds. Quarkus is already designed for assuming maximum build time initialization, since that's what GraalVM's original default was and that's what we've pushed towards. I don't really see that changing at all. Classes that already exist that could not be initialized at build time have already needed supplements provided by Quarkus and this has been shown to be a workable model. Combined with the possibility of standardization and the subsequent upstream alignment around standard mechanisms, even if the initial cost of complete support becomes *more* significant for Quarkus, the long term costs will trend towards zero or some minimum over time.  That minimum is likely to be lower without the class runtime init contagion problem.

**Comment by dmlloyd** [comment]
To show a rough "spitball" of what this might look like in terms of spec (both JVMS and JLS), imagine that could we introduce a mechanism like this:

```java

    // this method can be called during build time and its result will be cached forever after, including at run time; `const` here means "This is a ConstantDynamic method and its arguments must be constants"
    public static const long sum(long a, long b) { return a * b; }

    // same thing except this time, the result will only be cached until build time is over (or until a snapshot/restore happens, if that is a thing that will exist) 
    public static transient const long product(long a, long b) { return a * b; }

    // I don't want this called during build time!
    public static const Path findTempDir() {
        if (Runtime.getRuntime(). isBuildTime()) {
            throw new Error("This should not be called during build");
        }
    }
```

Note that I'm reusing some existing keywords here but that's just for the sake of illustration (like Brian Goetz often says "ignore the syntax!").

The necessary spec changes would be:

* Modify JVMS to provide a way to indicate that a `ConstantDynamic` is "transient"/not cached between JVM sessions
* Modify JLS to provide a declaration syntax for constant methods whose arguments must be constants (i.e. matching the requirements for `ConstantDynamic`)

Which is potentially much simpler for both specifier and user than having a complex definition for class initialization, or worse yet, using some kind of potentially surprising heuristics to try to figure out what classes can be safely initialized at build time.

**Comment by DanHeidinga** [comment]
> > To be clear, so was I. `<clinit>` is currently a mish-mash of state where some of it can be early initialized without issue while other bits (often in the same class!) can't be.
> > The challenge is that that isn't the language we have today. The "different mechanism" doesn't (yet) exist. And there are real cases that shouldn't be BTI that need to be expressed to make existing programs run in this environment.
> 
> It is not a requirement that we (or Leyden) need to try to make any existing software (including the JDK) be native-ready to _any_ degree on day one. In fact the Leyden project calls this out specifically. 

There's definitely going to be a transition period for the jdk core classes, libraries & frameworks, and finally applications to be native-ready.  So we agree that not everything has to be native-ready immediately but we equally have a problem if nothing is native-ready without massive amounts of work.  There's a balancing point here and we need to find the right set of tradeoffs.

> Given this, and due to the infectious nature of mixed class init, 

Yup

> and given that almost every native image benefit is derived directly or indirectly from front-loading class initialization to build time,

Having an initial heap of the pre-loaded classes, and their initialized state is a big part of the benefit, but let's not forget that having pre-compiled code for all the methods is another critical component here as that avoids the long warmup due to phase transitions from interpreted -> c1 -> c2.  We need to be careful to not overstate the benefit of pre-init.

> _and_ based on our experiences with Quarkus, I don't think it is nearly as expedient or useful as it would appear to be to compromise on this for the sake of compatibility with existing code. 

There's an 80/20 rule happening here.  Pre-init is really important but isn't the only thing.  There's a lot of actions that application developers could be taking to speed up their startup today - use of Hotspot's AppCDS / OpenJ9's sharedclasses, refactoring their existing code to lazily init statics, adopt Graal's native image, using CRIU to checkpoint/restore - and we're only starting to see some of these things in practice.  Many complaints about startup / footprint fail the "Mom test" (Mom doesn't want to hurt your failings so she validates your idea) [1].

The "best solution" that leaves the ecosystem behind will struggle to gain adoption.  We a solution that helps developers retain their existing code bases while bringing them along the adoption curve.

[1] http://momtestbook.com/

> Artificially striving for any level compatibility (or for that matter introducing any other ancillary requirements of this kind) can easily negatively impact the quality of the end result (for example by potentially complicating the definition of "class initialization" in the JVM spec). It's just not as nearly important to be maximally compatible as it is for native images (or the broader "static images" given in the Leyden proposal) to justify their limitations by maximizing material benefit compared to a dynamic JVM in terms of image size and startup speed.
> 

We're both looking at the same (technical) problem from different application ends - I'd like maximum applicability to the broader ecosystem even if it means a slightly bigger image / some runtime init.  In part, because I think we'll need some runtime init anyway for the all the reasons that the Graal folks have had to provide the various levels of {buildtime, runtime, and rerun}.

> In the short term, using pragmatic/hacky substitutions would be more than adequate to bridge between worlds as we've shown already with Quarkus and as has been shown by the efforts of upstream library developers and users. It is certainly the case that pre-standardization, our mechanism and GraalVM's will not likely be compatible unless we went to extraordinary measures to make it so, which in context seems like a waste of effort that could be spent on pushing forward a standardizable solution.
> 

I think we agree here - Substitutions are a dead-end long term though that might be a short-term pragmatic approach.  Long term we need a build-time class file load hook (CFLH) for agents to opt-into the build process just as they do with the existing  runtime CFLH.

> In the medium to long term, libraries would be migrating towards the standard approach to having native-friendly run time initialization (which for most libraries, even specialized ones like Netty, would likely only be a quite small overall percentage of code). This migration would be helped along by community contributions (certainly including substantial contributions from us).
> 

+1.

> > A class can be loaded but not initialized so I'm unclear on the "Being unable to even _refer_ to a run-time-initialized class during build time" problem.
> 
> > The case we've run into that I'm aware of is using special constants to indicate the platform / target environment we're building for - and I agree that's been a nuisance. A special case of treating those classes, assuming it's a small number, as "must be BTI" is certainly something we can do. The general case feels too far for the existing applications we're starting with.
> 
> The applications we are starting with, beyond "hello world" of course, are the set of applications which Quarkus builds. Quarkus is already designed for assuming maximum build time initialization, since that's what GraalVM's original default was and that's what we've pushed towards. I don't really see that changing at all. Classes that already exist that could not be initialized at build time have already needed supplements provided by Quarkus and this has been shown to be a workable model. 

Even with Quarkus, I understood (maybe incorrectly?) that runtime init was still required for some cases.  There's lots that can be done - and the Quarkus team has certainly pushed this amazing distance - but runtime init can't be completely avoided.

> Combined with the possibility of standardization and the subsequent upstream alignment around standard mechanisms, even if the initial cost of complete support becomes _more_ significant for Quarkus, the long term costs will trend towards zero or some minimum over time. That minimum is likely to be lower without the class runtime init contagion problem.

I think this overstates the case - runtime init will need to be part of the final model.  A path that supports both, and the ability to move from one to the other, is a better long term approach.

**Comment by DanHeidinga** [comment]
I think the problem - and the required spec work - is incredibly more complex than this rough "spitball" covers.  I'll respond to the parts here and try to explain after why I think this misses the complexity.

> To show a rough "spitball" of what this might look like in terms of spec (both JVMS and JLS), imagine that could we introduce a mechanism like this:
> 
> ```java
>     // This method can be called during build time *or* run time and its result will be cached forever after, including at run time.
>     // It does not need to be specified whether the constants are resolved lazily or eagerly during static init build; both should probably be allowed!
>     // `const` here means "This is a ConstantDynamic method and its arguments must be constants"
>     public static const long sum(long a, long b) { return a * b; }
>     //            ^^^^^
> ```

ConstantDynamic is a constantpool entry type.  It names a bootstrap method that is used to create the constant and that constant can then be used by `ldc` or as an argument to other bootstrap methods.  The existing semantics don't make sense in the context of a method that takes parameters. 

Proceeding on the assumption this is meant to be a constexpr type method.  So the next question is: to lower this to bytecode, what the translation be?  All callers of `sum(a, b)` would translate to one of two things:
* when `a` & `b` are known to be buildtime constants, to an `ldc ConstantDynamic:sumBSM:{a, b}`?
* when `a` & `b` aren't known to be buildtime constants, to a call to `sum(a,b)`?  I'm struggling to map this syntax to how condy works.
 
Even for the first case, I'm not sure why condy would be necessary.  If we're looking to mark methods as applicable to buildtime execution, then javac could do the `a*b` and replace the `sum(a,b)` call with the result.

If the intention here is to add `constexpr` to the java language, then condy has nothing to do with it.  We'd need very different semantics for how this would work and what would be considered "constant".  For an example of prior investigations into related ideas, see Brian Goetz's JVMLS talk "below the fold" that shows some javac constant folding and dead code elimination using Condy and Constantable: https://www.youtube.com/watch?v=iSEjlLFCS3E&list=PLX8CzqL3ArzVnxC6PYxMlngEMv3W1pIkn&index=2

> ```
>     // same thing except this time, the result will only be cached until build time is over (or until a snapshot/restore happens, if that is a thing that will exist) 
>     public static transient const long product(long a, long b) { return a * b; }
>     //            ^^^^^^^^^ ^^^^^
>  ```

I'm not clear on the semantics of this operation given the questions about the previous one.  Can you describe the bytecode you see being generated for this and the lifecycle it requires?

> ```
>     // I don't want this called during build time!
>     public static const Path findTempDir() {
>         if (Runtime.getRuntime().isBuildTime()) {
>             throw new Error("This should not be called during build");
>         }
>     }
> ```

As we run more code at buildtime, the code may very well need to know where is the temp dir so it can store its intermediate steps.  Or it needs to call something - maybe `createTempFile()` - that calls this method, now we can't do anything with temp files at build time.  This feels a bit like the classic "what color is your function?" [2] problem in a slightly different context.

[2] https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/

> 
> Note that I'm reusing some existing keywords here but that's just for the sake of illustration (like Brian Goetz often says "ignore the syntax!").
> 
> The necessary spec changes would be:
> 
> * Modify JVMS to provide a way to indicate that a `CONSTANT_Dynamic` is "transient"/not cached between JVM sessions, or perhaps something like `ACC_TRANSIENT` to be placed on the method declaration
> * Modify JVMS to indicate that a method is a constant method (`ACC_CONSTANT` or something)
> * Modify JLS to provide a declaration syntax for constant methods whose arguments must be constants (i.e. matching the requirements for `CONSTANT_Dynamic`)
> * _Edit - forgot about this one:_ Modify the JDK spec to provide methods to detect whether we're currently building a static image, running a static image, or running in a dynamic JVM (this part is already essentially agreed upon by Mark Reinhold though maybe not through this exact/specific mechanism)
> 
> Which is potentially much simpler for both specifier and user than having a complex definition for class initialization, or worse yet, using some kind of potentially surprising heuristics to try to figure out what classes can be safely initialized at build time.

This doesn't even begin to come close to the spec changes.  First, we need to nail down the **semantics** of the operations and then we can start to figure out how to express them in bytecode that both a native image and a dynamic JVM can run.

And the minute we move out of the land of built-in primitive types we fall straight into the existing `<clinit>` model and the soupy mish-mash of operations that it encodes.

Let's say we want to do something more complex as some kind of `constexpr` operation.  Say, build some kind of data structure at compile time:
```java

static Map<String, Object> map;
static {
  map =  new HashMap<>();
  map.add("Foo", List.of("bar", "baz");
}
```

Every class referenced in this example -`Map, HashMap, String, Object, List` - all have to be compatible with buildtime initialization.  As do all their super classes and interfaces they implement.  As do all the classes referenced from their `<clinit>` methods as well.  As do all the classes references by `HashMap::<init>()V`, `List::of()`, etc.

Any and all of these classes can have `<clinit>` methods that get triggered in this simple static initializer.  The answer of "throw it all away" doesn't work.  We need to find a way to refactor existing code to express developer intent about when things are initialized and to work with the `<clinit>` model to support that.

Maybe the lazy static final fields draft jep [3] and some new kind of `constexpr` give us the tools to hollow out `<clinit>` but we'll still need to figure out the model - and the spec - for how these things interact.  There's too many invariants being expressed in code that depend on the `<clinit>` model to break it for BTI.

[3] https://openjdk.java.net/jeps/8209964

**Comment by dmlloyd** [comment]
> > and given that almost every native image benefit is derived directly or indirectly from front-loading class initialization to build time,
> 
> Having an initial heap of the pre-loaded classes, and their initialized state is a big part of the benefit, but let's not forget that having pre-compiled code for all the methods is another critical component here as that avoids the long warmup due to phase transitions from interpreted -> c1 -> c2. We need to be careful to not overstate the benefit of pre-init.

It is not overstated. Without preinitialization there isn't evidence that there is enough benefit to static images to bother with standardization. The benefits have to be worth the limitations, and if static images just amount to avoiding compiler warmup then jaotc would already have been a success instead of having been retired.

> > _and_ based on our experiences with Quarkus, I don't think it is nearly as expedient or useful as it would appear to be to compromise on this for the sake of compatibility with existing code.
> 
> There's an 80/20 rule happening here. Pre-init is really important but isn't the only thing. There's a lot of actions that application developers could be taking to speed up their startup today - use of Hotspot's AppCDS / OpenJ9's sharedclasses, refactoring their existing code to lazily init statics, adopt Graal's native image, using CRIU to checkpoint/restore - and we're only starting to see some of these things in practice. Many complaints about startup / footprint fail the "Mom test" (Mom doesn't want to hurt your failings so she validates your idea) [1].

But most users aren't doing any of those things. They're complicated and require *more* run time infrastructure, not less.

If I tell someone who wants a binary executable for their application that they can get similar gains by using shared classes and CRIU, they're almost all going to say "no, I don't want all that stuff, I want a binary executable".  That's the goal of this project and of Leyden standardization and none of those other technologies really matter so much in that context.

> The "best solution" that leaves the ecosystem behind will struggle to gain adoption. We a solution that helps developers retain their existing code bases while bringing them along the adoption curve.

Our experience has shown that this is truly a false dichotomy. Even specifying build time initialization by default, Quarkus has not really had a significant problem here. In the cases where it is a problem, there are pragmatic solutions possible. There is never a need to compromise the long term goal and cost/benefit for short term gains in this way. It's a concession that is dramatically outsized compared to what has been observed as the actual problem.  Having a higher bar for static image compatibility imposes a one time cost on the ecosystem, whereas maintaining the compromises necessary for a low bar will incur ongoing costs forever. The one time cost is always the better deal.

> > Artificially striving for any level compatibility (or for that matter introducing any other ancillary requirements of this kind) can easily negatively impact the quality of the end result (for example by potentially complicating the definition of "class initialization" in the JVM spec). It's just not as nearly important to be maximally compatible as it is for native images (or the broader "static images" given in the Leyden proposal) to justify their limitations by maximizing material benefit compared to a dynamic JVM in terms of image size and startup speed.
> 
> We're both looking at the same (technical) problem from different application ends - I'd like maximum applicability to the broader ecosystem even if it means a slightly bigger image / some runtime init. In part, because I think we'll need some runtime init anyway for the all the reasons that the Graal folks have had to provide the various levels of {buildtime, runtime, and rerun}.

This is where we differ. The user definitely needs some way to express work that needs to be done or redone at run time. Class initialization is not the right mechanism for this. It is the mechanism that GraalVM had access to, and that is presumably why they used it - there was no other language option.  A new mechanism is necessary.

> > The applications we are starting with, beyond "hello world" of course, are the set of applications which Quarkus builds. Quarkus is already designed for assuming maximum build time initialization, since that's what GraalVM's original default was and that's what we've pushed towards. I don't really see that changing at all. Classes that already exist that could not be initialized at build time have already needed supplements provided by Quarkus and this has been shown to be a workable model.
> 
> Even with Quarkus, I understood (maybe incorrectly?) that runtime init was still required for some cases. There's lots that can be done - and the Quarkus team has certainly pushed this amazing distance - but runtime init can't be completely avoided.

No, runtime init is never _required_ by Quarkus. What is required is some way to do some things at run time, and with GraalVM, class init is the mechanism to do that. If there were a different mechanism, then Quarkus would use that instead. A common "pattern" that represents the difficulty of relying on static initialization as a solution involves creating a secondary static class which contains the static data that need reinitialization. This class is separately marked as run time initialized. Then, special annotations are used to redirect run-time accesses to _some_ of the primary class to the secondary class fields instead.  This is just one example of the ugliness of relying on multi-stage class initialization.

Class initialization is the wrong granularity for this task, and the wrong dependency model as well.

> > Combined with the possibility of standardization and the subsequent upstream alignment around standard mechanisms, even if the initial cost of complete support becomes _more_ significant for Quarkus, the long term costs will trend towards zero or some minimum over time. That minimum is likely to be lower without the class runtime init contagion problem.
> 
> I think this overstates the case - runtime init will need to be part of the final model. A path that supports both, and the ability to move from one to the other, is a better long term approach.

Any long term approach that relies on adding a complex mechanism that has to be maintained indefinitely is a worse long term approach by any measure. The ongoing costs can't compete with even a fairly high one-time cost - which build time init is definitely _not_, as has been shown empirically, even _without_ a nicer solution to run time initialization tasks.

> ConstantDynamic is a constantpool entry type. It names a bootstrap method that is used to create the constant and that constant can then be used by `ldc` or as an argument to other bootstrap methods. The existing semantics don't make sense in the context of a method that takes parameters.

The bootstrap method takes parameters which are in turn constant values. In this sketch, calling a constant method would only be allowed with constant arguments.

> Proceeding on the assumption this is meant to be a constexpr type method. So the next question is: to lower this to bytecode, what the translation be? All callers of `sum(a, b)` would translate to one of two things:
> 
> * when `a` & `b` are known to be buildtime constants, to an `ldc ConstantDynamic:sumBSM:{a, b}`?

Sure, it could be done that way.  Really there's no reason why it couldn't be a plain invokestatic; the semantics are what matter.  I'm simply trying to pick the fastest and easiest possible implementation path.

> * when `a` & `b` aren't known to be buildtime constants, to a call to `sum(a,b)`?  I'm struggling to map this syntax to how condy works.

No, it'd be a compilation error.

> Even for the first case, I'm not sure why condy would be necessary. If we're looking to mark methods as applicable to buildtime execution, then javac could do the `a*b` and replace the `sum(a,b)` call with the result.

In fact it's not necessary, it's just one way to solve the problem, being a tool that happens to be available and has a good match to the desired behavior.

> > ```
> >     // same thing except this time, the result will only be cached until build time is over (or until a snapshot/restore happens, if that is a thing that will exist) 
> >     public static transient const long product(long a, long b) { return a * b; }
> >     //            ^^^^^^^^^ ^^^^^
> > ```
> 
> I'm not clear on the semantics of this operation given the questions about the previous one. Can you describe the bytecode you see being generated for this and the lifecycle it requires?

The method itself could be `ACC_TRANSIENT | ACC_CONST | ACC_PUBLIC | ACC_STATIC`. Semantically, in its simplest form, each call site (or perhaps unique combination of constant arguments) would be replaced by the result of the corresponding method call which is computed by the first such caller.  If the JVM session ends (build time completion or perhaps snapshot taken) then any result of a transient const function is discarded.  This is similar to condy but of course condy doesn't have to be the mechanism by which this is implemented; it just reflects a popular idea.  It could be an `invokestatic` bytecode, with specific verifier rules to prevent arguments which are not in turn constants or the result of a const method call.  But a bootstrap method fits well here in terms of rules and results and would, in my estimation, require less JVMS revision (forbidding calling const methods via `invokestatic` would probably be it).

In this context `const` means "constant, one-time". And `transient` means "per-session".

Overall this is not unlike the lazy constants proposal you referenced below, but I'll point out that that JEP proposal also does not include bytecode specifics other than alluding to condy a few times.  Also the extra nominal dependency orderings imposed by that proposal are somewhat counterintuitive (but not incompatible) if that idea were expanded to include `transient`.

> As we run more code at buildtime, the code may very well need to know where is the temp dir so it can store its intermediate steps. Or it needs to call something - maybe `createTempFile()` - that calls this method, now we can't do anything with temp files at build time. This feels a bit like the classic "what color is your function?" [2] problem in a slightly different context.

This solution is designed specifically to avoid this problem at the language level. But it is an unavoidable problem in the end - some work is always only going to be correct during build time, some work is always only going to be correct during run time, and some work has to be repeated for each session. The key is to minimize dependency interactions between these cases.

By preventing these interactions at the language level - you either have something that is done once globally, or once per session - we don't have to worry about it there since it's impossible to incorrectly define this interaction. By letting the user throw the exception, they have complete control over when they want these things to run and what happens when they are run at an unexpected time.

> This doesn't even begin to come close to the spec changes. First, we need to nail down the **semantics** of the operations and then we can start to figure out how to express them in bytecode that both a native image and a dynamic JVM can run.
> 
> And the minute we move out of the land of built-in primitive types we fall straight into the existing `<clinit>` model and the soupy mish-mash of operations that it encodes.

I don't see why this would be true.  The point is that each operation is fine grained.

> Let's say we want to do something more complex as some kind of `constexpr` operation. Say, build some kind of data structure at compile time:
> 
> ```java
> static Map<String, Object> map;
> static {
>   map =  new HashMap<>();
>   map.add("Foo", List.of("bar", "baz");
> }
> ```
> 
> Every class referenced in this example -`Map, HashMap, String, Object, List` - all have to be compatible with buildtime initialization. As do all their super classes and interfaces they implement. As do all the classes referenced from their `<clinit>` methods as well. As do all the classes references by `HashMap::<init>()V`, `List::of()`, etc.

Yes, this is exactly the intention. Any library which is not compatible with build time initialization is not static image ready, and that is OK. Quarkus and other containers can continue to use substitutions for this case, as well as contributing upstream patches to make libraries static image ready.

> Maybe the lazy static final fields draft jep [3] and some new kind of `constexpr` give us the tools to hollow out `<clinit>` but we'll still need to figure out the model - and the spec - for how these things interact. There's too many invariants being expressed in code that depend on the `<clinit>` model to break it for BTI.

It's not breaking the clinit model at - in fact it's the opposite, it's preserving it _exactly_. Classes have to be initialized before they are used; there is no time that is before build time. However once you *push* initialization later to run time, now you've got the red/blue function problem right in your lap. Run time init classes can depend on build time init classes, but not the other way around, and experience shows that these dependencies can sneak in more easily than one would think, thanks to the coarse granularity of init which results in hungry dependency graphs.

Even if Leyden at a spec level allows for a spectrum of init behaviors by implementations, then upstream developers still need to assume that some implementations will require build time init and must plan accordingly by using non-clinit mechanisms for things that _must_ be done at run time, or risk incompatibility with more optimized implementations. And of course if they have things that _must_ be done at build time then they need to be able to express that intent; if clinit is not guaranteed to be run at build time for static init then it cannot be used for that either, but this is something that is otherwise much harder to express because a static image compiler with a "flexible" initialization policy would necessarily need to be told if something has to be build time initialized using a keyword of some sort. It would be problematic rely on the user throwing an exception, like you can in the aggressive build time model, because such an exception would only appear when the program is run.

The language mechanism has to have a consistent, predictable, and repeatable behavior model.  Applying a heuristic to move initialization is neither consistent nor predictable nor repeatable.  A flat build time initialization is all of these.  Mixing build and run time clinit is none of these.  A model that allows one-time actions _and_ one-time-per-session actions is all of these.

**Comment by DanHeidinga** [comment]
> > Having an initial heap of the pre-loaded classes, and their initialized state is a big part of the benefit, but let's not forget that having pre-compiled code for all the methods is another critical component here as that avoids the long warmup due to phase transitions from interpreted -> c1 -> c2. We need to be careful to not overstate the benefit of pre-init.
> 
> It is not overstated. Without preinitialization there isn't evidence that there is enough benefit to static images to bother with standardization. 

Pre-init is important.  We both agree on that.  But not every class initializer is the same.  Some, as you allude to below, are used to group statics that should initialized as late as possible and hold state that's specific to the execution environment - things like the user temp dir, the environment variables, etc - are all things that a user would reasonably want to be stored in a static final field (**express intent**) and would thus have in a static initializer.  Or used to initialize a thread pool or ... there's an infinite set of things users can do in a static initializer and some of them really benefit from being done at build time.  Others 100% can't be.  And others don't matter either way.

We should be careful to examine the other use cases and why they matter to users before we discard them an unimportant.  I've been surprised by how excited many folks have been to have a single executable for their java application even if it provided the same performance (or less) then they got with a dynamic jvm.

> The benefits have to be worth the limitations, and if static images just amount to avoiding compiler warmup then jaotc would already have been a success instead of having been retired.

The benefits have to be worth the limitations - agreed.  But we should be clear on **which** benefits, and **who** benefits, and who pays the **costs**.  If supporting pre-initialization requires a lot of different ways of thinking about how code is written, and requires testing in a native env, and results in a worse experience (perf / debugging / clarity / etc) on the dynamic jvm, then it will be hard for library writers to want to opt in and that makes it even harder for app developers to opt in.

The benefits require an ecosystem lift and need to be framed to not leave that ecosystem behind.  Or cause hard rifts in it.

There's a lot I could say about the approach `jaotc` took that I would do differently (and we did differently in OpenJ9) but I don't think it adds to the point.  We both agree a solution in this space needs to bring clear benefits.  In my mind that's to both developers on the JVM and on static images.  - Not so different from what Qaurkus does in supporting both.

> If I tell someone who wants a binary executable for their application that they can get similar gains by using shared classes and CRIU, they're almost all going to say "no, I don't want all that stuff, I want a binary executable". That's the goal of this project and of Leyden standardization and none of those other technologies really matter so much in that context.

This speaks to my point above - many developers want an executable because they want a single executable.  It's not the startup / footprint benefits they want.

When I mention the other technologies that help startup / warmup, we should note that many people who complain about startup speed fall into the "complainers bucket" from the "mom test" book in that they complain but have taken no action to improve their situation.  In the book, that means they won't buy into your startup.  Here, it means many of them won't make the changes to adopt the new tech.

> 
> > The "best solution" that leaves the ecosystem behind will struggle to gain adoption. We a solution that helps developers retain their existing code bases while bringing them along the adoption curve.
> 
> Our experience has shown that this is truly a false dichotomy. Even specifying build time initialization by default, Quarkus has not really had a significant problem here. In the cases where it is a problem, there are pragmatic solutions possible. 

Quarkus has targeted greenfield deployments:  My understanding from the talks I've seen is that Quarkus is (was?) primarily focused on new applications.  These are applications that would have an easier time adapting to new restrictions.

> There is never a need to compromise the long term goal and cost/benefit for short term gains in this way. It's a concession that is dramatically outsized compared to what has been observed as the actual problem. Having a higher bar for static image compatibility imposes a one time cost on the ecosystem, whereas maintaining the compromises necessary for a low bar will incur ongoing costs forever. The one time cost is always the better deal.

It may be a one time cost but it's also a bar preventing adoption.  We don't want to end up in the same state as modules where it's hard to adopt them until your dependencies have adopted them.  

> > We're both looking at the same (technical) problem from different application ends - I'd like maximum applicability to the broader ecosystem even if it means a slightly bigger image / some runtime init. In part, because I think we'll need some runtime init anyway for the all the reasons that the Graal folks have had to provide the various levels of {buildtime, runtime, and rerun}.
> 
> This is where we differ. The user definitely needs some way to express work that needs to be done or redone at run time. Class initialization is not the right mechanism for this. It is the mechanism that GraalVM had access to, and that is presumably why they used it - there was no other language option. A new mechanism is necessary.

We need new mechanisms that let users express their intent.  Whether that new mechanism is tied to class initialization or not is still an open question.  And we'll need to specify how that mechanism interacts with class initialization.

If we start to put today's static fields - and the graph of objects accessible from them - into buckets, we have (at least):
* a build time bucket ("as early as possible")
* a runtime bucket ("as late as possible")
* a don't care bucket ("do it whenever, provided it's initialized by first use")

Do you see other buckets?

One I'm hesitant to include is a "redo bucket" for things that have been to be initialized to use but really need to be initialized at runtime.  A potential example of this is the default fork join pool used by parallel streams which might get initialized as a side-effect of the "build time bucket" but really wants to bind tightly to the number of cpus in the runtime environment.

> No, runtime init is never _required_ by Quarkus. What is required is some way to do some things at run time, and with GraalVM, class init is the mechanism to do that. If there were a different mechanism, then Quarkus would use that instead. A common "pattern" that represents the difficulty of relying on static initialization as a solution involves creating a secondary static class which contains the static data that need reinitialization. This class is separately marked as run time initialized. Then, special annotations are used to redirect run-time accesses to _some_ of the primary class to the secondary class fields instead. This is just one example of the ugliness of relying on multi-stage class initialization.

This feels a bit like a semantic game.  There is static state in the application that needs to be initialized at runtime.  Yes, it's done with `<clinit>` today and maybe there's a better mechanism we can develop, but it doesn't change the fact that some things will always need runtime initialization.

As to the initialization-on-demand-holder (IODH) pattern, from what I've looked at, applications can get quite a ways by grouping their statics into different IODH buckets.  Pulling further on this string is worth while as it may help to identify the buckets (from above) and help us nail down the required semantics.

> 
> Class initialization is the wrong granularity for this task, and the wrong dependency model as well.
> 
> > > Combined with the possibility of standardization and the subsequent upstream alignment around standard mechanisms, even if the initial cost of complete support becomes _more_ significant for Quarkus, the long term costs will trend towards zero or some minimum over time. That minimum is likely to be lower without the class runtime init contagion problem.
> > 
> > 
> > I think this overstates the case - runtime init will need to be part of the final model. A path that supports both, and the ability to move from one to the other, is a better long term approach.
> 
> Any long term approach that relies on adding a complex mechanism that has to be maintained indefinitely is a worse long term approach by any measure. The ongoing costs can't compete with even a fairly high one-time cost - which build time init is definitely _not_, as has been shown empirically, even _without_ a nicer solution to run time initialization tasks.

It seems like we're talking past each other.  Class initialization today entangles all static state into a single `<clinit>` method.  It's the only thing we have in today's language.  

We both agree some of that state is best initialized at build time and some of that state will need to be initialized at runtime.  We would be further ahead if we focused on the semantics of those different buckets and thought about how to recognize which things have to go into each bucket.  Right now, we have some very rough heuristics but it would be better if we could expand and specify those rules.

> 
> > ConstantDynamic is a constantpool entry type. It names a bootstrap method that is used to create the constant and that constant can then be used by `ldc` or as an argument to other bootstrap methods. The existing semantics don't make sense in the context of a method that takes parameters.
> 
> The bootstrap method takes parameters which are in turn constant values. In this sketch, calling a constant method would only be allowed with constant arguments.

Ok.  Constant arguments - as the "Below the Fold" talk I linked above shows - isn't immediately clear once you step away from directly using primitive values.

Is `sum(a, b)` a constant express?  It's easy to say yes when you see:
``` java
sum(1, 1)
```
but 
```java
int a = 5;
int b = 7;
sum(a,b)
```
is less clear, and so is
```java
int a = 5;
int b;
if (System.getProperty("os.name").equals("WIN")) {
b = 7;
} else {
b = 8;
}
sum(a,b);
```
For a translation strategy ("javac") to handle this, it needs to be able to do the constant propagation for you - something it doesn't do by design today apart from the most simple cases.


> 
> > Proceeding on the assumption this is meant to be a constexpr type method. So the next question is: to lower this to bytecode, what the translation be? All callers of `sum(a, b)` would translate to one of two things:
> > 
> > * when `a` & `b` are known to be buildtime constants, to an `ldc ConstantDynamic:sumBSM:{a, b}`?
> 
> Sure, it could be done that way. Really there's no reason why it couldn't be a plain invokestatic; the semantics are what matter. I'm simply trying to pick the fastest and easiest possible implementation path.

We 100% agree the semantics matter most.  Trying to use ConstantDynamic as an implementation shortcut is obscuring the semantics.  Let's ignore the implementation to start and really clearly state the semantics.  Then we figure out the right way to translate that into classfiles, and for both native images and the dynamic vm to execute those semantics.

> > * when `a` & `b` aren't known to be buildtime constants, to a call to `sum(a,b)`?  I'm struggling to map this syntax to how condy works.
> 
> No, it'd be a compilation error.
> 
> > Even for the first case, I'm not sure why condy would be necessary. If we're looking to mark methods as applicable to buildtime execution, then javac could do the `a*b` and replace the `sum(a,b)` call with the result.
> 
> In fact it's not necessary, it's just one way to solve the problem, being a tool that happens to be available and has a good match to the desired behavior.

Condy has great semantics for dealing with moving individual static fields out of the `<clinit>` method and giving them their own "as late as possible" semantics.  It's not a great tool for modeling "as early as possible" semantics.

> 
> > > ```
> > >     // same thing except this time, the result will only be cached until build time is over (or until a snapshot/restore happens, if that is a thing that will exist) 
> > >     public static transient const long product(long a, long b) { return a * b; }
> > >     //            ^^^^^^^^^ ^^^^^
> > > ```
> > 
> > 
> > I'm not clear on the semantics of this operation given the questions about the previous one. Can you describe the bytecode you see being generated for this and the lifecycle it requires?
> 
> The method itself could be `ACC_TRANSIENT | ACC_CONST | ACC_PUBLIC | ACC_STATIC`. Semantically, in its simplest form, each call site (or perhaps unique combination of constant arguments) would be replaced by the result of the corresponding method call which is computed by the first such caller. 

So this is the "redo" bucket I was hesitant to suggest above.  A value that's computed and then recomputed later.  

> If the JVM session ends (build time completion or perhaps snapshot taken) then any result of a transient const function is discarded. This is similar to condy but of course condy doesn't have to be the mechanism by which this is implemented; it just reflects a popular idea. It could be an `invokestatic` bytecode, with specific verifier rules to prevent arguments which are not in turn constants or the result of a const method call. 

The verifier is the most expensive, and difficult, tool to use to enforce something like this.  It doesn't track values, it tracks types to ensure that a program is type-safe (to verifier rules which are different than the language rules).  

> But a bootstrap method fits well here in terms of rules and results and would, in my estimation, require less JVMS revision (forbidding calling const methods via `invokestatic` would probably be it).

Let's step back from implementation.  I think it's leading us to take shortcuts.  Let's focus on the semantics as drive by the use cases we envision.

> > As we run more code at buildtime, the code may very well need to know where is the temp dir so it can store its intermediate steps. Or it needs to call something - maybe `createTempFile()` - that calls this method, now we can't do anything with temp files at build time. This feels a bit like the classic "what color is your function?" [2] problem in a slightly different context.
> 
> This solution is designed specifically to avoid this problem at the language level. But it is an unavoidable problem in the end - some work is always only going to be correct during build time, some work is always only going to be correct during run time, and some work has to be repeated for each session. The key is to minimize dependency interactions between these cases.

Ok.  We agree here.  This is the "as early as possible", "as late as possible", and "redo" buckets.  Minimizing dependency between them is good and will require us to be clear on what constitutes a dependency, which direction those dependencies can point in (redo can point to early but not late, late can point to early and redo, early can only point to early?), and the semantics for how they deal with legacy `<clinit>`.
 
> By preventing these interactions at the language level - you either have something that is done once globally, or once per session - we don't have to worry about it there since it's impossible to incorrectly define this interaction. By letting the user throw the exception, they have complete control over when they want these things to run and what happens when they are run at an unexpected time.

Globally correct invariants are good.  I guess that makes the `Runtime.getRuntime().isBuildTime()` the moral equivalent of an ArrayStoreException thrown when a validly typed program hits a case that can't be detected at source code compile time?  

> It's not breaking the clinit model at - in fact it's the opposite, it's preserving it _exactly_. Classes have to be initialized before they are used; there is no time that is before build time. However once you _push_ initialization later to run time, now you've got the red/blue function problem right in your lap. Run time init classes can depend on build time init classes, but not the other way around, and experience shows that these dependencies can sneak in more easily than one would think, thanks to the coarse granularity of init which results in hungry dependency graphs.

This sounds very similar to what I said above.   So I think we're in the same ballpark.  

> 
> Even if Leyden at a spec level allows for a spectrum of init behaviors by implementations, then upstream developers still need to assume that some implementations will require build time init and must plan accordingly by using non-clinit mechanisms for things that _must_ be done at run time, or risk incompatibility with more optimized implementations. 

+1 to having a mechanism that supports the "a runtime bucket ("as late as possible")".  I'm less comfortable saying this can't / won't be a `<clinit>`.

> And of course if they have things that _must_ be done at build time then they need to be able to express that intent; if clinit is not guaranteed to be run at build time for static init then it cannot be used for that either, but this is something that is otherwise much harder to express because a static image compiler with a "flexible" initialization policy would necessarily need to be told if something has to be build time initialized using a keyword of some sort. 

The challenge here is that there is one set of source code for most libraries.  (And libraries, as always, are where the challenges are more so than the application code as they have to support a broader set of use cases).  Up-leveling the coloured function problem to coloured libraries ("must be build time" / "must be runtime") is still failure.

We need semantics that can be executed by both a dynamic jvm and by a static image.

> It would be problematic rely on the user throwing an exception, like you can in the aggressive build time model, because such an exception would only appear when the program is run.

Yup.  We should exhaust every other option before introducing such a mechanism.

> The language mechanism has to have a consistent, predictable, and repeatable behavior model. Applying a heuristic to move initialization is neither consistent nor predictable nor repeatable. A flat build time initialization is all of these. Mixing build and run time clinit is none of these. A model that allows one-time actions _and_ one-time-per-session actions is all of these.

To progress these, in qbicc we make these failed BTI `<clinit>`'s into exception throws if we add sufficient logging to make it easy to know which `<clinit>`s failed and critically **why**.  This will help us to form the buckets mentioned above.

In parallel, we should look at how to hollow out `<clinit>` in those cases and what the semantics we need are.

**Comment by dmlloyd** [comment]
The original issue is fixed with #760, and the ongoing init discussion is happening in #764.
